### ISSUE TITLE
vmap implementation for torch.linalg.eigh (aten::linalg_eigh)

### ISSUE THREAD

**Issue by deoliveirajoshua** [pytorch/pytorch#119522]
### 🚀 The feature, motivation and pitch

I'm working with a batch of Hessian tensors of shape [B, K, N, N], where each element of the batch is 3rd-order Hessian tensor [K, N, N]. While torch.linalg.eigh already has a batching feature for tensors of the shape [K, N, N],  I would like it if there was a technique for computing the eigen decomposition of all BxK hermitian matrices while still preserving the structure of the tensor. This would make it additionally friendlier / seemless for performing hessian analysis, as vmap is already implemented for torch.func.hessian, and already can produce Hessian tensors of shape [B, K, N, N] for a batch of B instances on a K-dimensional vector-valued function, ie: $f: \mathbb{R}^N \to \mathbb{R}^k $. 

### Alternatives

Tensors with shape [B, K, N, N] can be collapsed into tensors of shape [B*K, N, N], run torch.linalg.eigh, then reshaped back to [B, K, N, N]. This, however, requires that you first look up the len() of dims 0, 1, and 2 reshapes. 

Other than the API consistency previously mentioned, I think vmap'ing this feature will provide a speed-up, which would be greatly useful as vmap(torch.func.hessian) can require significant resources already when dealing with large batches / highly-parameterized functions.

### Additional context

![Screenshot 2024-02-08 204235](https://github.com/pytorch/pytorch/assets/37802360/b45f46c6-f2be-4452-acec-17c19d4ce0b0)

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano @zou3519

**Comment by lezcano** [comment]
@guilhermeleobas this should be easy.

**Comment by guilhermeleobas** [comment]
@deoliveirajoshua, which version of PyTorch are you running? `torch.linalg.eigh` decomposition was implemented in https://github.com/pytorch/pytorch/pull/110640 and was landed on PyTorch 2.2. On previous versions I can see the warning but not on the most recent one.

Reproducer:
```python
import torch

def fn(x):
    return torch.ops.aten.linalg_eigh(x)

def wrapper(x):
    return vmap(fn)(x)

x = torch.randn(4, 3, 3)
print(wrapper(x))
```
